src/disasm_wrapper.c

```diff
@@ -0,0 +1,291 @@
+/*
+ * Copyright 2016-2017, Intel Corporation
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ *       notice, this list of conditions and the following disclaimer.
+ *
+ *     * Redistributions in binary form must reproduce the above copyright
+ *       notice, this list of conditions and the following disclaimer in
+ *       the documentation and/or other materials provided with the
+ *       distribution.
+ *
+ *     * Neither the name of the copyright holder nor the names of its
+ *       contributors may be used to endorse or promote products derived
+ *       from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/*
+ * disasm_wrapper.c -- connecting the interceptor code
+ * to the disassembler code from the capstone project.
+ *
+ * See:
+ * http://www.capstone-engine.org/lang_c.html
+ */
+
+#include "intercept.h"
+#include "intercept_util.h"
+#include "disasm_wrapper.h"
+
+#include <assert.h>
+#include <string.h>
+#include <syscall.h>
+#include "capstone_wrapper.h"
+
+struct intercept_disasm_context {
+	csh handle;
+	cs_insn *insn;
+	const unsigned char *begin;
+	const unsigned char *end;
+};
+
+/*
+ * nop_vsnprintf - A dummy function, serving as a callback called by
+ * the capstone implementation. The syscall_intercept library never makes
+ * any use of string representation of instructions, but there seems to no
+ * trivial way to use dissassemble using capstone without it spending time
+ * on printing syscalls. This seems to be the most that can be done in
+ * this regard i.e. providing capstone with nop implementation of vsnprintf.
+ */
+static int
+nop_vsnprintf()
+{
+	return 0;
+}
+
+/*
+ * intercept_disasm_init -- should be called before disassembling a region of
+ * code. The context created contains the context capstone needs ( or generally
```

Comment:
Yes. I think have an old branch somewhere, using NASM with interface in ```disasm_wrapper.h```
Other disassemblers should probably be easy to use as well.